core-rust/state-manager/src/state_manager.rs

```diff
@@ -712,123 +711,132 @@ where
-                    .expect("prepared transaction not committable");
-                let hash_structures_diff = &commit.hash_structures_diff;
-                let state_hash_tree_diff = hash_structures_diff.state_hash_tree_diff.clone();
-                let transaction_tree_slice =
-                    hash_structures_diff.transaction_tree_diff.slice.clone();
-                let receipt_tree_slice = hash_structures_diff.receipt_tree_diff.slice.clone();
-                let local_receipt = commit.local_receipt.clone();
-                let database_updates = commit.database_updates.clone();
-                (
-                    state_hash_tree_diff,
-                    transaction_tree_slice,
-                    receipt_tree_slice,
-                    local_receipt,
-                    database_updates,
-                )
-            };
-
-            Self::check_epoch_proof_match(
-                commit_ledger_header,
-                series_executor.next_epoch(),
-                i == (commit_transactions_len - 1),
-            )?;
+            let commit = series_executor
+                .execute(config_type, &validated, "prepared")
+                .expect("cannot execute transaction to be committed");
 
             if let Some(intent_hash) = validated.intent_hash_if_user() {
                 intent_hashes.push(intent_hash);
             }
 
-            substate_store_update.apply(database_updates);
-            state_tree_update.add(series_executor.latest_state_version(), state_hash_tree_diff);
-            transaction_tree_slice_merger.append(transaction_tree_slice);
-            receipt_tree_slice_merger.append(receipt_tree_slice);
+            substate_store_update.apply(commit.database_updates.clone());
+            let hash_structures_diff = commit.hash_structures_diff;
+            state_tree_update.add(
+                series_executor.latest_state_version(),
+                hash_structures_diff.state_hash_tree_diff,
+            );
+            transaction_tree_slice_merger.append(hash_structures_diff.transaction_tree_diff.slice);
+            receipt_tree_slice_merger.append(hash_structures_diff.receipt_tree_diff.slice);
 
-            result_receipts.push(local_receipt.clone());
+            result_receipts.push(commit.local_receipt.clone());
 
             committed_transaction_bundles.push(CommittedTransactionBundle {
                 state_version: series_executor.latest_state_version(),
                 raw,
-                receipt: local_receipt,
+                receipt: commit.local_receipt,
                 identifiers: CommittedTransactionIdentifiers {
                     payload: validated.create_identifiers(),
                     resultant_ledger_hashes: *series_executor.latest_ledger_hashes(),
                 },
             });
         }
 
-        let commit_ledger_hashes = &commit_ledger_header.hashes;
+        if series_executor.next_epoch() != commit_ledger_header.next_epoch.as_ref() {
```

Comment:
_(note from author)_

I realized it's enough to check this only once, after all transactions.
(the `series_executor.state_tracker` makes sure that an epoch change, if any, happens at the last transaction of the batch)